### ISSUE TITLE
workqueue tests don't run with session-scoped DFK

### ISSUE THREAD

**Issue by benclifford** [Parsl/parsl#1228]
The workqueue executor does not always seem to work running multiple tasks with a DFK instance.

This has come up from work on PR #1210 which rearranged tests to use a single DFK for a whole session instead of a new DFK for each `test_*` function; but I have also seen this fail when running a command line parsl test. The present test driver contains a workqueue specific bodge to work around this.

This command usually fails on a test:

```pytest parsl -k "not cleannet and not noci" --config parsl/tests/configs/workqueue_ex.py --cov=parsl --cov-append --cov-report= --random-order```

This command, which is what is used in CI now and which starts up one DFK for each test function, does not:

```
pytest parsl -k "not cleannet and not noci" --config parsl/tests/configs/workqueue_ex.py --cov=parsl --cov-append --cov-report= --random-order --bodge-dfk-per-test
```

I haven't dug much into this but sometimes I've seen the code hit this codepath in the WQ executor:

```
                            else:
                                reason += "unable to process wrapper script failure with status = {}".format(status)
```

with `status = -1`.

**Comment by dthain** [comment]
@tjdasso adding you to this conversation.  Next thing to look at after your current WQ PR gets finalized.

**Comment by benclifford** [comment]
@tjdasso @dthain let me know if you need more info about what is actually going on here on the parsl side

**Comment by tjdasso** [comment]
@benclifford could you give a general rundown of what goes on when the tests run, and what the current WQ workaround does to provide a temporary fix?

**Comment by benclifford** [comment]
@tjdasso what happened before recent rearrangements in #1210 was that each test file got an individual `pytest` invocation, in a for-loop in .travis.yaml.

This incurred a bunch of startup time and the test suite took many minutes to run.

PR #1210 changed things so that pytest created a single DFK at the start of the test run, ran all the tests using that DFK, and then shut down that DFK.

This means that the whole stack of DFK and everything underneath it (including in the case of this issue, the workqueue executor and workqueue worker executable) get a different workload: instead of getting a very small number of tasks (often, only one), the entire set of pytest tasks goes to the same DFK/executor.

This doesn't work for all configurations ... but it should!

Notably it doesn't work for exex (at least for me) which is issue #1229 even though exex is tested for a small number of tasks elsewhere in the test suite.

And it doesn't work for workqueue (at least for me).

So the bodge that I put in to work around this is the `pytest` parameter: ` --bodge-dfk-per-test` which you should find in .travis.yml only on the workqueue test, and not on the other pytest invocations.

That changes the way in which DFKs (and the other stuff in the stack below it) is initialised: rather than creating one DFK at the start of the pytest run and shutting it down at the end of the pytest run, instead a new DFK is created around each test (as before).

To recreate this:

Check that the workqueue pytest command line in `.travis.yml` passes for you:

```
pytest parsl -k "not cleannet" --config parsl/tests/configs/workqueue_ex.py --cov=parsl --cov-append --cov-report= --random-order --bodge-dfk-per-test
```

Next, run the same command, but without `--bodge-dfk-per-test`. I expect it to fail.

**Comment by tjdasso** [comment]
Running it locally with or without the `--bodge-dfk-per-test` flag still caused the non-skipped tests to fail for me. I have the up-to-date parsl code, but perhaps I have a different WQ setup on my personal computer.

**Comment by benclifford** [comment]
Can you run any parsl+WQ on your own computer? if so, then the fact that the tests are failing are a start to investigating problems, even if they are different to the ones in CI...

**Comment by tjdasso** [comment]
I just ran some basic Parsl/WQ tests from previous testing and they were working. I'll look into the `pytest` failures that are happening on my local computer

**Comment by benclifford** [comment]
@tjdasso look in .travis.yml if you want to see the other commands related to workqueue. start around here: https://github.com/Parsl/parsl/blob/472273cf7d13e7301b13806fbeb01f52d6b960ab/.travis.yml#L54

**Comment by tjdasso** [comment]
@benclifford my shell session didn't have `export PARSL_TESTING=True` when I originally ran the test with `--bodge_dfk_per_test`, but now all the tests have passed when running locally. Currently running without the bodge now

**Comment by tjdasso** [comment]
I just ran without the `--bodge_dfk_per_test` and I got similar results to what you mentioned. Almost all of the tests passed, but a few tests failed at `test_bash_apps/test_error_codes.py` and at `test_bash_apps/test_stdout.py`, with the same wrapper script `status = -1`. However, it also returned with `WORK_QUEUE_RESULT_STATUS = 1`, which signifies that there was a missing input file (actually, this part of the WQ executor code needs updating because it skips the error handling for the `WORK_QUEUE_RESULT_STATUS` if the Parsl status is also not 0). Do these tests use specific input files that might not be present locally?

**Comment by tjdasso** [comment]
Actually, it's failing on `@bash_app`s, so I am assuming that `status = -1` code is a result of running a `@bash_app` that eventually returns -1, which is then passed up through the `workqueue_worker.py` script and the -1 is handled by the executor.

**Comment by benclifford** [comment]
the main thing I'm unclear of and haven't dug into is why it works one way and not the other way -- my guess is that running multiple tests against one workqueue executor means that it's doing something  for a particular test that then breaks things for another test... because the individual tests themselves pass fine. so I'm assuming there's an interaction between different tests at the workqueue executor level that isn't clear here.

**Comment by tjdasso** [comment]
I think it goes back to the way the executor handles `@bash_app`s, which wasn't as rigorously tested on the Work Queue end. Yes, it seems as though running multiple tests within the same instance is causing a side effect that is in turn affecting the results of the next test. I'll look in the `executor.py` code more closely to see if I can find what's going on

**Comment by benclifford** [comment]
For some narrowing down, I see that this always fails for me on my laptop - just running one set of stdout related tests:

pytest parsl/tests/test_bash_apps/test_stdout.py -k "not cleannet" --config parsl/tests/configs/workqueue_ex.py  --random-order --maxfail=1

but succeeds with --bodge-dfk-per-test

If I run the workqueue worker with --debug, I see workqueue missing file errors for the last test that runs. It looks like something to do with finding/naming of t.err, which is a stderr output file output by several tests in that test run.

```
2019/11/21 20:05:15.20 work_queue_worker[7991] wq: rx from master: ./cctools-monitor --no-pprint --with-output-files=cctools-monitor -dall -o cctools-monitor.debug --with-time-series -L 'cores: 8' -L 'memory: 15845' -L 'disk: 13140' -V 'task_id: 1' -V 'category: default' --sh "python3 workqueue_worker.py -i task_0_function_data -o task_0_function_result -r t.err:t.err"
2019/11/21 20:05:15.20 work_queue_worker[7991] wq: rx from master: outfile file-4-d1951b527dbeca8bededa29bd5cafd66-t.err t.err 0
2019/11/21 20:05:15.20 work_queue_worker[7991] wq: started process 8022: ./cctools-monitor --no-pprint --with-output-files=cctools-monitor -dall -o cctools-monitor.debug --with-time-series -L 'cores: 8' -L 'memory: 15845' -L 'disk: 13140' -V 'task_id: 1' -V 'category: default' --sh "python3 workqueue_worker.py -i task_0_function_data -o task_0_function_result -r t.err:t.err"
2019/11/21 20:05:16.30 work_queue_worker[7991] wq: moving output file from /tmp/worker-1000-7991/t.1/t.err to cache/file-4-d1951b527dbeca8bededa29bd5cafd66-t.err
2019/11/21 20:05:16.30 work_queue_worker[7991] wq: could not rename output file /tmp/worker-1000-7991/t.1/t.err to cache/file-4-d1951b527dbeca8bededa29bd5cafd66-t.err: No such file or directory
2019/11/21 20:05:16.30 work_queue_worker[7991] wq: could not copy output file /tmp/worker-1000-7991/t.1/t.err to cache/file-4-d1951b527dbeca8bededa29bd5cafd66-t.err: No such file or directory
2019/11/21 20:05:16.30 work_queue_worker[7991] wq: rx from master: get file-4-d1951b527dbeca8bededa29bd5cafd66-t.err 1
2019/11/21 20:05:16.30 work_queue_worker[7991] wq: tx to master: missing file-4-d1951b527dbeca8bededa29bd5cafd66-t.err 2
2019/11/21 20:05:16.30 work_queue_worker[7991] wq: rx from master: unlink file-4-d1951b527dbeca8bededa29bd5cafd66-t.err
2019/11/21 20:05:16.34 work_queue_worker[7991] wq: rx from master: unlink file-13-d1951b527dbeca8bededa29bd5cafd66-t.err
```

That log talks about file-4-*-t.err mostly, but also mentions file-13-*.t.err

**Comment by dthain** [comment]
We are teasing apart several intersecting problems in this issue.
Here is what we know so far:

Problem 1 - The WQ Executor translates certain error codes coming back from WQ into custom exceptions that start with "Work Queue System Failure".  The test suite doesn't like this b/c it is expecting some other specific exception type, such BadStdStreamFile.  Two possible solutions: a) Make the test suite less specific about exception types.  b) Make the WQ executor generate exactly the same exception type as expected.

Problem 2: Some tasks runs successfully when we do not reuse the WQ executor across multiple DFK invocations.  Hypothesis: Some files passed to tasks are incorrectly marked as "cacheable" but in the test suite, the same input file name is used multiple times with different contents, leading to inconsistencies.  Test: Disable caching of files in the WQ Executor and see if the problem goes away.  If it does, a caching problem is certain.  If not, something else.

**Comment by dthain** [comment]
Note to @tjdasso: enable the WQ log on the master side, so we can see any decisions made by the master that happen before the task even gets to a worker.  For example, if the master discovers that an input file to a task does not exist, then it doesn't send the task to any worker at all, but just marks it as failed with status INPUT_MISSING.

**Comment by dthain** [comment]
@btovar suggests that the mechanics of starting a worker in the test will run better if the application creates the Factory object to start workers locally.  That way, the test framework is only running one top-level process.

**Comment by benclifford** [comment]
PR #1498 looks at adding provider based functionality to the workqueue executor. I think if that was in place, then parsl testing could start up workqueue workers in the same way that it starts up high throughput executor workers: parsl starts, decides it needs workers, starts some, shuts them down at the end. This would avoid the need for anything test-specific wrt starting workers. I'm not sure how that fits in with the general model of starting up workqueue workers locally/on HPC resources.

I'm unclear how this relates to this issue #1228 though.

**Comment by btovar** [comment]
@benclifford Yes, I think my comment is more appropriate to pr #1498.  

I am able to reproduce  your results. With --bodge-dfk-per-test it works, without it it does not. Investigating...

**Comment by btovar** [comment]
It does not seem to be caching related, as the very first task fails to produce the required output.

**Comment by btovar** [comment]
It is not caching, but some name clashing. For example, in:
https://github.com/Parsl/parsl/blob/4709b305b78047b3f58cd5a943507081f12744bd/parsl/tests/test_bash_apps/test_stdout.py#L46
there is `stderr='t.err'` which it is a name assigned to multiple tasks. As a test, I changed to something like `stderr='t.err.' + str(time.time())`, and now 9/10 tests succeed. I am investigating what that last error is about.

Right now the workqueue executor has a flat namespace. It seems that it needs to create a local temporary directory per task sent to it to avoid these name clashes. I can work on this if this sounds like a good idea.

**Comment by benclifford** [comment]
fixed by #1705 #1763
